src/session.ts

```diff
@@ -0,0 +1,278 @@
+
+import utils = require('./utils')
+import kernel = require('./kernel')
+
+import Signal = phosphor.core.Signal;
+import emit = phosphor.core.emit;
+import IAjaxSuccess = utils.IAjaxSuccess;
+import IAjaxError = utils.IAjaxError;
+
+
+/**
+ * Notebook Identification specification.
+ */
+export
+interface INotebookId {
+  path: string;
+};
+
+
+/**
+ * Session Identification specification.
+ */
+export
+interface ISessionId {
+  id: string;
+  notebook: INotebookId;
+  kernel: kernel.IKernelId;
+};
+
+
+/**
+ * Session initialization options.
+ */
+export
+interface ISessionOptions {
+  notebookPath: string;
+  kernelName: string;
+  baseUrl: string;
+  wsUrl: string;
+};
+
+
+/**
+ * Session object for accessing the session REST api. The session
+ * should be used to start kernels and then shut them down -- for
+ * all other operations, the kernel object should be used.
+ **/
+export 
+class Session {
+
+  static statusChanged = new Signal<Session, string>();
+
+  /**
+   * GET /api/sessions
+   *
+   * Get a list of the current sessions.
+   */
+  static list(sessionServiceUrl: string): Promise<ISessionId[]> {
+    return utils.ajaxRequest(sessionServiceUrl, {
+      method: "GET",
+      dataType: "json"
+    }).then((success: IAjaxSuccess): ISessionId[] => {
+      if (success.xhr.status === 200) {
+        if (!Array.isArray(success.data)) {
+          throw Error('Invalid Session list');
+        }
+        for (var i = 0; i < success.data.length; i++) {
+          validateSessionId(success.data[i]);
+        }
+        return success.data;
+      }
+      throw Error('Invalid Status: ' + success.xhr.status);
+    });
+  }
+
+  /**
+   * Construct a new session.
+   */
+  constructor(options: ISessionOptions) {
+    this._notebookModel = {
+      path: options.notebookPath
+    };
+    this._kernelModel = {
+      id: null,
+      name: options.kernelName
+    };
+
+    this._baseUrl = options.baseUrl;
+    this._wsUrl = options.wsUrl;
+    this._sessionServiceUrl = utils.urlJoinEncode(this._baseUrl, 
+                                                  'api/sessions');
+  }
+
+  /**
+   * Get the session kernel object.
+  */
+  get kernel() : kernel.Kernel {
+    return this._kernel;
+  }
+
+  /**
+   * POST /api/sessions
+   *
+   * Start a new session. This function can only executed once.
+   */
+  start(): Promise<void> {
+    return utils.ajaxRequest(this._sessionServiceUrl, {
+      method: "POST",
+      dataType: "json",
+      data: JSON.stringify(this._model),
+      contentType: 'application/json'
+    }).then((success: IAjaxSuccess) => {
+      if (success.xhr.status !== 201) {
+        throw Error('Invalid response');
+      }
+      validateSessionId(success.data);
+      this._updateModel(success.data);
+      if (this._kernel) {
+        this._kernel.name = this._kernelModel.name;
+      } else {
+        var kernelServiceUrl = utils.urlPathJoin(this._baseUrl, "api/kernels");
```

Comment:
Sessions take the base url and intrinsicially knows the path to add for the kernel service URL.  Kernels require the instantiator to know the kernel service URL.  The way session does it makes more sense to me - there's no reason why the session.ts file should have to know the kernel api url.  Can we change the kernel class to automatically know to add the 'api/kernels' url to the base url, and make the kernel constructor just take the base url?